README.md
# clj-knn

Implementation of k-nearest neighbors in Clojure.

### Regression

```
(def X
  [[1 6]
   [2 4]
   [3 7]
   [6 8]
   [7 1]
   [8 4]])

(def y [7 8 16 44 50 68])
(def data (make-data X y))

(assert (= 8 (regression 1 euclidean data [4 2])))
(assert (= 42 (regression 3 euclidean data [4 2])))
(assert (= 29 (regression 1 manhattan data [4 2])))
(assert (= 71/2 (regression 3 manhattan data [4 2])))
(assert (= 171/5 (regression-2 3 manhattan data [4 2])))
(assert (<= 37 (regression-2 3 euclidean data [4 2]) 38))
```

### Classification

```
(def X
  [[1 6]
   [2 4]
   [3 7]
   [6 8]
   [7 1]
   [8 4]])

(def y-class [:slow :slow :fast :fast :fast :light-speed])
(def data-class (make-data X y-class))

(assert (= :slow (classify 1 euclidean data-class [4 2])))
(assert (= :slow (classify 3 euclidean data-class [4 2])))
(assert (= :fast (classify 3 manhattan data-class [5 6])))
(assert (= :fast (classify 3 euclidean data-class [5 6])))
```

## knn

k-NN is a type of instance-based learning, or lazy learning, where the function is only approximated locally and all computation is deferred until classification. The k-NN algorithm is among the simplest of all machine learning algorithms.

The following distance metrics are provided:
* Euclidean
* Taxicab/Manhattan

The k-nn implementation is simple to extend by providing your own distance metrics (Should meet the [Mercer's Conditions](https://en.wikipedia.org/wiki/Mercer%27s_condition)).

It currently supports the following prediction policies:
* Classification
  * Mode
* Regression
  * Mean
  * LWR (Locally Weighted Regression)

## License

Copyright © 2016 Arthur Caillau

Distributed under the Eclipse Public License either version 1.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-03-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-03-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/clj-knn/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clj-knn/compare/0.1.0...0.1.1


test/clj_knn/core_test.clj
(ns clj-knn.core-test
  (:require [clojure.test :refer :all]
            [clj-knn.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clj_knn/utils.clj
(ns clj-knn.utils)

(defn square [x]
  {:pre [(number? x)]
   :post [(number? %)]}
  (* x x))

(defn abs [x]
  {:pre [(number? x)]
   :post [(number? %)]}
  (max x (- x)))

(defn mean [xs]
  {:pre [(seq xs) (every? number? xs)]
   :post [(number? %)]}
  (/ (reduce + xs) (count xs)))

(assert (= 1 (mean [0 1 2])))
(assert (= 0 (mean [0 0 0])))

(defn mode [xs]
  {:post [(or (get (set xs) %) (nil? %))]}
  (let [freqs (frequencies xs)]
    (first (reduce (fn [[k-max v-max] [k v]] (if (> v v-max) [k v] [k-max v-max]))
                   (first freqs)
                   (rest freqs)))))

(assert (= :a (mode [:a :a :a :b :c])))
(assert (= :d (mode (shuffle [:a :a :a :b :c :c :d :d :d :d]))))
(assert (not (mode [])))

(defn inverse [x]
  (if (zero? x)
    (Integer/MAX_VALUE)
    (/ 1 x)))

(assert (= 1 (inverse 1)))
(assert (= 1/2 (inverse 2)))

(defn lwr [xs weights]
  (/ (reduce + (map * xs weights))
     (reduce + weights)))

(assert (= 1 (lwr [1 2] [1 0])))
(assert (= 3/2 (lwr [1 2] [1 1])))

(defn make-data [X y]
  (mapv vector X y))

(defn ^:private take-at-least-aux [n [x & xs :as coll] acc]
  (cond
    (not (seq coll)) acc
    (and (<= n 0) (not= x (last acc))) acc
    :else (recur (dec n) xs (conj acc x))))

(assert (= [1 1] (take-at-least-aux 2 [1 1 2] [])))
(assert (= [1 1] (take-at-least-aux 1 [1 1 2] [])))
(assert (= [] (take-at-least-aux 0 [1 1 2] [])))

(defn take-at-least [n coll]
  (take-at-least-aux n coll []))


src/clj_knn/core.clj
(ns clj-knn.core
  (:require [clj-knn.utils :refer [inverse lwr mode mean take-at-least make-data]]
            [clj-knn.distances :refer [euclidean manhattan]]))

; TODO: Not very efficient - Use a min-heap O(k.lg(N)) instead of O (N.lg(N))
(defn predict [prediction-policy]
  "Prediction policy takes in a sequence of values
   [[[X1 y1] distance], ... [[Xk yk] distancek]] and returns a y

   Output: Returns a function that takes in the following keys:

   - k: Integer
   - distance-metric: similarity function - Mercer's Conditions
   - data: all the data for the algo to be able to predict - lazy learning
   - datapoint: a new datapoint to make a prediction on

   - Output: the prediction on the new-datapoint"
  (fn [k distance-metric data datapoint]
    (let [distances (map (comp (partial distance-metric datapoint) first) data)
          k (->> distances sort (take-at-least k) count)]
      (->> distances
           (map vector data)
           (sort-by second)
           (take k)
           prediction-policy))))

(def ys (partial map (comp second first)))
(def distances (partial map second))

(def classify (predict (comp mode ys)))
(def regression (predict (comp mean ys)))
; Locally Weighted Regression
(def regression-2 (predict (comp (partial apply lwr)
                                 (juxt ys
                                       (comp (partial map inverse)
                                             distances)))))

; Test Cases
; ------------
(def X
  [[1 6]
   [2 4]
   [3 7]
   [6 8]
   [7 1]
   [8 4]])

(def y [7 8 16 44 50 68])
(def data (make-data X y))

(assert (= 8 (regression 1 euclidean data [4 2])))
(assert (= 42 (regression 3 euclidean data [4 2])))
(assert (= 29 (regression 1 manhattan data [4 2])))
(assert (= 71/2 (regression 3 manhattan data [4 2])))
(assert (= 171/5 (regression-2 3 manhattan data [4 2])))
(assert (<= 37 (regression-2 3 euclidean data [4 2]) 38))

(def y-class [:slow :slow :fast :fast :fast :light-speed])
(def data-class (make-data X y-class))

(assert (= :slow (classify 1 euclidean data-class [4 2])))
(assert (= :slow (classify 3 euclidean data-class [4 2])))
(assert (= :fast (classify 3 manhattan data-class [5 6])))
(assert (= :fast (classify 3 euclidean data-class [5 6])))
